### PR TITLE
Testing accelerate CFD interface 

### LINKED ISSUE #1626: [TASK] Interface for Accelerate CFD
--------
Issue Description
--------

**Is your feature request related to a problem? Please describe.**
Create RAVEN interface for Accelerate CFD (ACFD-RAVEN API) to automate the process. (https://github.com/IllinoisRocstar/AccelerateCFD_CE)

The ACFD-RAVEN API aims to enhance the industrial market value of a fast-running CFD ROM (Reduced Order Model) software, AccelerateCFD, under development by Illinois Rocstar LLC, pursuing two key capabilities to achieve the goal. First, the overall process of AccelerateCFD’s ROM creation, starting from high-fidelity CFD simulations, can be optimized by incorporating RAVEN’s optimization capability through the new interface. Second, the optimization target can be extended from the AccelerateCFD’s internal nhyperparameters such as number of bases to the overall execution process,  meaning that an advanced (automated) ROM construction process / workflow (i.e., new API) can be developed within the new working framework. Another expected benefit from this development is that the practical utility of the AccelerateCFD (or its surrogate models) can be enhanced by utilizing the RAVEN’s capability of uncertainty quantification (through input parameters perturbation) by suggesting the associated model uncertainties or design margins.  

**Describe the solution you'd like**

**Describe alternatives you've considered**

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [ ] 1. If the issue is a defect, is the defect fixed?
- [ ] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [ ] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [ ] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [ ] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by wangcj05**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Based on Andrea's branch alfoa_ga_acc_cdf, see PR #1574.

Create ACFD-RAVEN API  to enhance the industrial market value of a fast-running CFD ROM (Reduced Order Model) software, AccelerateCFD, under development by Illinois Rocstar LLC, pursuing two key capabilities to achieve the goal. 
close #1626 

Testing the modifications 

##### What are the significant changes in functionality due to this change request?

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

**Comment by wangcj05** [comment]
@aalfonsi I have cleaned up your branch, and made several modifications to make all tests pass. If you have time, please take a look. 

@joshua-cogliati-inl do you have time to review this PR? There are several modifications made by Andrea in JobHandler, MPISimulationMode, and ray related modification, and I really want you to take a look at these changes. 

Currently, the test includes a lot of files, and each file is pretty large, and we need to find a way to reduce size of the test.

File: framework/CustomModes/MPISimulationMode.py [hunk 1]
```diff
@@ -55,6 +55,13 @@ def modifyInfo(self, runInfoDict):
       @ In, runInfoDict, dict, the original runInfo
       @ Out, newRunInfo, dict, of modified values
     """
+    magicSyntax = False
```
Comment (aalfonsi) [inline framework/CustomModes/MPISimulationMode.py:58]: I think these modifications are in the RAY branch.... I would toss them out

File: framework/JobHandler.py [hunk 2]
```diff
@@ -181,17 +181,49 @@ def __initializeRay(self):
     if self.runInfoDict['internalParallel']:
       ## Check if the list of unique nodes is present and, in case, initialize the
       servers = None
+      sys.path.append(self.runInfoDict['WorkingDir'])
+      if 'UPDATE_PYTHONPATH' in self.runInfoDict:
+        sys.path.extend([p.strip() for p in self.runInfoDict['UPDATE_PYTHONPATH'].split(":")])
+
+      self.rayInstanciatedOutside = False
```
Comment (aalfonsi) [inline framework/JobHandler.py:188]: see previous comment

**Review by aalfonsi — CHANGES_REQUESTED** [review]
All the modifications for the jobhandler and cluster interaction are in the RAY branch..i would remove them from here

File: framework/CustomModes/MPISimulationMode.py (hunk 1 above)
Comment (wangcj05) [reply framework/CustomModes/MPISimulationMode.py:58]: Thanks Andrea, I have resolved the conflict of ray branch, but there are several issues happened to the PR. We will try to resolve and merge the ray pr, and then we will rebase this branch.

**Comment by joshua-cogliati-inl** [comment]
There is no description of what this change does and no issue associated with it.

File: framework/CustomModes/MPISimulationMode.py (hunk 1 above)
Comment (joshua-cogliati-inl) [reply framework/CustomModes/MPISimulationMode.py:58]: The MPISimulationMode changes should be removed.

Comment (wangcj05) [reply framework/CustomModes/MPISimulationMode.py]: removed

File: framework/JobHandler.py (hunk 2 above)
Comment (wangcj05) [reply framework/JobHandler.py]: removed

**Comment by wangcj05** [comment]
@joshua-cogliati-inl I have added the issue and reverted several unneeded changes.

**Comment by wangcj05** [comment]
@joshua-cogliati-inl This PR is ready for you to review.

**Comment by wangcj05** [comment]
@joshua-cogliati-inl @aalfonsi I think this PR is ready for you to review.

**Review by aalfonsi — APPROVED** [review]

**Review by joshua-cogliati-inl — COMMENTED** [review]
Generally looks reasonable, but there are some minor things to fix.

File: framework/CodeInterfaces/AccelerateCFD/AccelerateCFD.py [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+    for child in xmlNode:
+      if child.tag == 'outputLocations':
+        coords = child.find("coordinates")
+        if coords is None:
+          raise IOError("Sub-node <coordinates> not found in node <outputLocations>!")
+        coordinates = coords.text.split()
+        cds = []
+        for coord in coordinates:
+          if "(" not in coord or ")" not in coord:
+            raise IOError("<coordinates> must be inputted with the foloowing format (x1,y1,z1) ")
```
[the remark this answers is not in the document]
Comment (wangcj05) [reply framework/CodeInterfaces/AccelerateCFD/AccelerateCFD.py]: fixed

File: framework/CodeInterfaces/AccelerateCFD/AccelerateCFD.py [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+          xyz = [c.strip() for c in coord.replace(")","").replace("(","").split(",")]
+          if len(xyz) != 3:
+            raise IOError("<coordinates> must be inputted with the foloowing format (x1,y1,z1) (x2,y2,z2) etc ")
```
[the remark this answers is not in the document]
Comment (wangcj05) [reply framework/CodeInterfaces/AccelerateCFD/AccelerateCFD.py]: fixed

File: framework/CodeInterfaces/AccelerateCFD/AccelerateCFD.py [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+    self.locations = {}
+    for child in xmlNode:
+      if child.tag == 'outputLocations':
+        coords = child.find("coordinates")
+        if coords is None:
+          raise IOError("Sub-node <coordinates> not found in node <outputLocations>!")
+        coordinates = coords.text.split()
+        cds = []
+        for coord in coordinates:
+          if "(" not in coord or ")" not in coord:
```
[the remark this answers is not in the document]
Comment (wangcj05) [reply framework/CodeInterfaces/AccelerateCFD/AccelerateCFD.py]: updated with your suggestion

File: framework/CodeInterfaces/AccelerateCFD/AccelerateCFD.py [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+    currentInputsToPerturb = self.findInps(currentInputFiles,"input")
+    originalInputs         = self.findInps(oriInputFiles,"input")
+    parser = GenericParser.GenericParser(currentInputsToPerturb)
+    parser.modifyInternalDictionary(**Kwargs)
+    parser.writeNewInput(currentInputsToPerturb,originalInputs)
+    return currentInputFiles
+
+  def readFoamFile(self, filename):
+    """
+      This method is aimed to read a Open Faom file for accelerated CFD
```
[the remark this answers is not in the document]
Comment (wangcj05) [reply framework/CodeInterfaces/AccelerateCFD/AccelerateCFD.py]: fixed

File: framework/CodeInterfaces/AccelerateCFD/AccelerateCFD.py [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+    numberOfProcs = len(resultFolders)
+    resultFolders.sort()
+    resultFolder =  resultFolders[0]
+    results = {}
+    resultingDirs = [os.path.join(resultFolder, o) for o in os.listdir(resultFolder)
+                        if os.path.isdir(os.path.join(resultFolder,o))]
+    resultingDirs.sort()
+    timeList = []
+    tsDirs =  resultingDirs
+    for ts in resultingDirs:
```
[the remark this answers is not in the document]
Comment (wangcj05) [reply framework/CodeInterfaces/AccelerateCFD/AccelerateCFD.py:276]: I have added some comments

File: framework/CodeInterfaces/AccelerateCFD/AccelerateCFD.py [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+import pandas as pd
+import xml.etree.ElementTree as ET
+import glob
+import math
+from sklearn import neighbors
+from OpenFoamPP import fieldParser
+from CodeInterfaceBaseClass import CodeInterfaceBase
+from GenericCodeInterface import GenericParser
+
+def findNearest(array, value):
```
Comment (mandd) [inline framework/CodeInterfaces/AccelerateCFD/AccelerateCFD.py:32]: I think we have a similar method in mathUtils, if not, should we move it there?

File: tests/framework/CodeInterfaceTests/AcceleratedCFD/test_perturb_acdf.xml [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+      volVectorField-max_max_max-y,
+      volVectorField-max_max_max-z,
+      volVectorField-middle_middle_middle-x,
+      volVectorField-middle_middle_middle-y,
+      volVectorField-middle_middle_middle-z
+      </Output>
+    </PointSet>
+    <HistorySet name="AcceleratedCFDHistorySet">
+      <Input>velMods,scalMods</Input>
+      <Output>time,volVectorField-0_0_0_0_0_0-x,
```
Comment (mandd) [inline tests/framework/CodeInterfaceTests/AcceleratedCFD/test_perturb_acdf.xml:88]: should time be the pivot param?

File: framework/CodeInterfaces/AccelerateCFD/AccelerateCFD.py (hunk 8 above)
Comment (wangcj05) [reply framework/CodeInterfaces/AccelerateCFD/AccelerateCFD.py]: I have updated the mathUtils method to handle this.

File: tests/framework/CodeInterfaceTests/AcceleratedCFD/test_perturb_acdf.xml (hunk 9 above)
Comment (wangcj05) [reply tests/framework/CodeInterfaceTests/AcceleratedCFD/test_perturb_acdf.xml:88]: yes, our default pivot param is 'time', and our code will try to remove index from the output.

**Comment by wangcj05** [comment]
@mandd I have resolved your comments

**Review by joshua-cogliati-inl — APPROVED** [review]
Reviewed and approve.

**Pull request merged**